fix(common): add `HttpParamsOptions` to the public api

rwlogel opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [X] Bugfix
- [x] Documentation content changes
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
The `HttpParamsOptions` is not documented or included in the public API even though it is a constructor argument of `HttpParams` which is in the public API.

Issue Number: #20276


## What is the new behavior?
`HttpParamsOptions` will be included in the public API and documented.

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

Commit:
fix(common): add `HttpParamsOptions` to the public api

The `HttpParamsOptions` was not documented or included in the public API even
though it is a constructor argument of `HttpParams` which is a part of the
public API. This commit adds the `HttpParamsOptions` into the exports, thus
making it a part of the public API.

Resolves #20276

IgorMinar reviewed (dismissed):
This is not a doc's change. Can you please change this comment to be a fix thank you

gkalpak commented:
For context: #20332 and #23015

rwlogel commented:
The reason I think we should expose the interface instead of making it inline as part of the `HttpParams` structure is because it makes extending `HttpParams` less complicated.  This is the situation where I wanted to use it to work around #11058:

```
export class PlusSafeQueryEncoder implements HttpParameterCodec {
  encodeKey(key: string): string {
    return encodeURIComponent(key);
  }

  encodeValue(value: string): string {
    return encodeURIComponent(value);
  }

  decodeKey(key: string): string {
    return decodeURIComponent(key);
  }

  decodeValue(value: string): string {
    return decodeURIComponent(value);
  }
}

export class PlusSafeHttpParams extends HttpParams {
  constructor(options?: HttpParamsOptions = {} as HttpParamsOptions) {
    super({ ...options, encoder: new PlusSafeQueryEncoder() })
  }
}
```

gkalpak commented:
Sounds reasonable to me.

You can always grab the type from the `HttpParams` constructor, but it still sounds reasonable to use an interface:
```ts
type MyHttpParamsOptions = ConstructorParameters<typeof HttpParams>[0];

jelbourn reviewed (approved):
LGTM

Reviewed-for: public-api

petebacondarwin reviewed (approved):
Reviewed-for: public-api

AndrewKushnir reviewed (approved):
The change LGTM, thanks @rwlogel.

Could you please rebase this PR and update the commit message body to include a bit more information (the `lint` check will most likely fail after a rebase since the commit message body should be at least 20 chars)? 

Here is a proposed commit message (feel free to use is as is or update it as you see fit):

```
fix(common): add `HttpParamsOptions` to the public api

The `HttpParamsOptions` was not documented or included in the public API even though it is a constructor argument of `HttpParams` which is a part of the public API. This commit adds the `HttpParamsOptions` into the exports, thus making it a part of the public API.

Resolves #20276.
```

I'm adding a "cleanup" label for now, which will be removed once PR is rebased and the commit message is updated.

Reviewed-for: public-api

atscott reviewed (approved):
Reviewed-for: public-api

rwlogel commented:
I have rebased and updated the commit message

> The change LGTM, thanks Ray Logel.
> 
> Could you please rebase this PR and update the commit message body to include a bit more information (the `lint` check will most likely fail after a rebase since the commit message body should be at least 20 chars)?
> 
> Here is a proposed commit message (feel free to use is as is or update it as you see fit):
> 
> ```
> fix(common): add `HttpParamsOptions` to the public api
> 
> The `HttpParamsOptions` was not documented or included in the public API even though it is a constructor argument of `HttpParams` which is a part of the public API. This commit adds the `HttpParamsOptions` into the exports, thus making it a part of the public API.
> 
> Resolves #20276.
> ```
> 
> I'm adding a "cleanup" label for now, which will be removed once PR is rebased and the commit message is updated.
> 
> Reviewed-for: public-api

mary-poppins commented:
You can preview 22be818 at https://pr35829-22be818.ngbuilds.io/.
You can preview c387233 at https://pr35829-c387233.ngbuilds.io/.
You can preview faeebb9 at https://pr35829-faeebb9.ngbuilds.io/.
You can preview 34628cc at https://pr35829-34628cc.ngbuilds.io/.
You can preview 51aad23 at https://pr35829-51aad23.ngbuilds.io/.

AndrewKushnir reviewed (approved):
Thanks for updating this PR!

I've also checked that the `HttpParamsOptions` is displayed correctly ([see here](https://pr35829-51aad23.ngbuilds.io/api/common/http/HttpParamsOptions)) and also started tests in Google's codebase ([internal-only link](http://test/OCL:323658414:BASE:323658458:1595972943638:3d69a9d6)).

Note: this PR should be ready for merge once all tests are completed and there is a final approval from the `fw-http` group (cc @alxhub).

Thank you.

alxhub reviewed (approved):
Reviewed-For: public-api

AndrewKushnir commented:
Started a [new presubmit](http://test/OCL:341671724:BASE:341671746:1605039447188:c0247423), will keep this thread updated.

AndrewKushnir commented:
FYI, presubmit is successful for the changes in this PR. Thank you.
